### ISSUE TITLE
How to use custom fetch options locally instead of globally

### ISSUE THREAD

**Issue by s-KaiNet** [SharePoint/PnP-JS-Core#499]
#### Category
[х] Enhancement
[х] Question

#### Version
[ latest ]

####  Question
I have a quite specific query, may be you can give some insights about how to proceed.     

`pnp-js` uses `pnp.setup` method in order to configure some global headers, fetch factory, etc. The problem is that this configuration is global to all `pnp-js` queries, but I need local setup. Let me clarify.     

For example I have a node.js web application and I want to use different types of access tokens in order to do some things in SharePoint. It is possible to construct user+app access token (in that case all operations in SharePoint will be on behalf of the user) or app only access token (used in pnp's NodeFetchClient).  
Now I want:  
```javascript
import {Web} from 'sp-pnp-js';
let userAccessToken = 'eyJ0eXAiO...';
let appOnlyToken: = 'eyJ0eXAiO....';
let anotherUserAccessToken = 'eyJ0eXAiO....';
pnp.sp.utility.sendEmail(emailProps, userAccessToken) // <-- sending email on behalf of the user
pnp.sp.utility.sendEmail(emailProps, anotherUserAccessToken ) // <-- sending on behalf of another user

new Web(url, userAccessToken).get().then(...) // <-- web with user access token
new Web(url, appOnlyAccessToken).get().then(...) // <-- web with app only access token etc.
```     
Since it's a web application, it's not possible to use global setup, because concurrent requests will use the same global config and the same headers (and access tokens) which is not valid. 
`pnp.setup()` allows me to have some global headers and no way to construct objects with predefined headers with different types of tokens or simply different headers.   

#### Possible solutions   
**Do not even read solution below, I've just tried to create a PR and I really don't like it - too many changes. For sure you will reject it.** :) 
~~I propose following changes in order to add support for such scenarios.~~    
~~Change `Queryable` constructor from~~    
```javascript
constructor(baseUrl: string | Queryable, path?: string)
```  
~~to~~
```javascript
constructor(baseUrl: string | Queryable, path?: string, options: FetchOptions = {})
```  
~~`options` will be merged later on in `toRequestContext` method. **Drawbacks**: this change affects all constructors for all classes derived from `Queryable` (around 30).   
Web constructor (as well as other objects derived from `Queryable`) will looks like:~~  
```javascript
new Web(url, undefined /* default path will be 'webs' */, { headers: {header: 'my header'}});
```    
~~Also `SPRest` class needs to be changed a bit in order to support something like this:~~   
```javascript
pnp.sp.useOptions({headers: etc..}).web.get().then(/* etc.. */)
```

I hope I was clear, please ask questions if any, if you want to see the PR in order to have better understanding I can go ahead and push one.     
What do you think on this issue? Or may be there is a better solution and I missed something?  

**UPD:** PR for proposed solution is extremely big and I'm pretty sure my solution is bad. The question becomes: do you have any recommendation how to workaround this issue?

**Comment by s-KaiNet** [comment]
Finally found a much more nicer way how to do that, welcome to PR #500 :)

**Comment by patrick-rodgers** [comment]
This was already possible:

1. You can use a new Web instance to create a new "base url" for a query
2. You can pass custom parameters to the get method that includes headers, but this is not something really exposed on the non-get methods, though the capability exists.

I will take a look at the PR - I am all for making things easier and let you know any feedback. Give me a few days, got some other stuff on my plate as well :).  Thanks!!

**Comment by s-KaiNet** [comment]
Thank you, Patrick, 
right, all Queryables have baseUrl as first param (this is something I'm utilizing in my PR) and all get methods accepts additional headers, but like you mentioned, the main issue is POST-able methods:  
```javascript
pnp.sp.web.applyTheme(...)
pnp.sp.web.fields.add(...)
etc...
```
no way to apply custom header.

**Comment by patrick-rodgers** [comment]
Made comments on the PR...I like where this is going.

**Comment by pedroivorbgrodrigues** [comment]
I need this PR! It's been a pain to make the sp-pnp-js work with node-sp-auth, i can only consume my sharepoint using node-sp-requester, but than its hard to work with list items and stuff.
Please release it!

**Comment by patrick-rodgers** [comment]
@pedroivorbgrodrigues - can you outline your scenario more clearly? I'd like to understand what is blocking you to ensure we can address it. Thanks!

**Comment by s-KaiNet** [comment]
@pedroivorbgrodrigues  did you see [node-pnp-js](https://github.com/s-KaiNet/node-pnp-js) or [sp-pnp-node](https://github.com/koltyakov/sp-pnp-node) (more advanced implementation)?

**Comment by patrick-rodgers** [comment]
Closing this as resolved with the merge of #500. Thanks @s-KaiNet for the great work on this one!

**Comment by pedroivorbgrodrigues** [comment]
Sorry for the late response, crazy times over here. I gave up on using the libs even though they worked, I managed to get the the list data from a different domain, but got a problem with the images links.

@s-KaiNet Yeah I looked them all up, I don't remember now which one made it working but I think 2 libs actually made it.

@patrick-rodgers My scenario was the following:
 I needed to expose a list as a blog to the internet, but didn't want to go throughout the SharePoint Web Application route. Them I created a nodejs express server and a html5 client, and the idea was: the client asks the server for the sp list, and the server consumes the SP api. I problem was I needed to authenticate, but I wanted the perks of the sp-pnp library (the odate consumption through chain methods). I managed to authenticate using @patrick-rodgers  solutions, but then I couldn't use it in the sp-pnp lib, didn't manage to anyway. When I managed to integrate booth solutions I ended up having a problem with the images inside a field in the list that were linking a document library in sharepoint.
I made a proxy for those images to pass thought the authenticated route and stream the image back to the cliente, but then I got some net::ERR CONTENT LENGTH MISMATCH and never manage to solve that.

Thank you guys for the great work! Will use this later on I hope. =]
